data-serving/reusable-data-service/reusable_data_service/controller/case_controller.py

```diff
@@ -109,20 +100,63 @@ def batch_upsert(self, body: dict):
         (created, updated) = (
             self.store.batch_upsert(usable_cases) if len(usable_cases) > 0 else (0, 0)
         )
-        status = 200 if len(errors) == 0 else 207
-        response = {"numCreated": created, "numUpdated": updated, "errors": errors}
-        return jsonify(response), status
+        return CaseUpsertOutcome(created, updated, errors)
+
+    def download(
+        self, format: str = "csv", filter: str = None, case_ids: List[str] = None
+    ):
+        """Download all cases matching the requested query, in the given format."""
+
+        if filter is not None and case_ids is not None:
+            raise PreconditionUnsatisfiedError(
+                "Do not supply both a filter and a list of IDs"
+            )
+
+        permitted_formats = ["csv", "tsv", "json"]
+        if format not in permitted_formats:
+            raise UnsupportedTypeError(f"Format must be one of {permitted_formats}")
+        # now we know the format is good, we can build method names using it
+        converter_method = f"to_{format}"
+        header_method = f"{format}_header"
+        separator_method = f"{format}_separator"
+        footer_method = f"{format}_footer"
+
+        if case_ids is not None:
+            case_iterator = self.store.identified_case_iterator(case_ids)
+            count = len(case_ids)
+        else:
+            predicate = CaseController.parse_filter(filter)
+            if predicate is None:
+                raise ValidationError("cannot understand query")
```

Comment:
Add the predicate and filter to this exception message?

Reply:
I didn't do it as the request already contains the query that wasn't understood (and the predicate is `None`), but can do.